### PR TITLE
Query units schema

### PR DESCRIPTION
**Pull request by christophermlawson**
This PR is is combining two older branches to deliver the functionality that allows the frontend to resolve Units obtained from an iModel in the backend.

### CODE REVIEW

**Review by pmconne — COMMENTED** [review]
Is it necessary to add yet another iModel (testUnits.bim) to the git repo? Can an existing iModel be used instead, or the tests be marked `#integration` and obtain the iModel from the hub?

**Comment by christophermlawson** [comment]
@pmconne I just need the Units schema imported one of the test bim files, so can I use one of the other files.  Suggestion as to which one?

**Comment by pmconne** [comment]
@christophermlawson I don't think it matters which one, but I'd go with test.bim. The presence of the units schema shouldn't impact any other tests unless there's one that's counting number of schemas or something, right?

**Comment by christophermlawson** [comment]
@pmconne yes, I would assume so.  I will try test.bim and see how it goes.

**Comment by christophermlawson** [comment]
@pmconne I'm using the testImodel.bim.  It already had the units schema.

**Review by pmconne — APPROVED** [review]

File: test-apps/ui-test-app/src/frontend/index.tsx
Comment (bsteinbk) [inline test-apps/ui-test-app/src/frontend/index.tsx:398]: @christophermlawson I am getting crash here when opening existing briefcase/snapshots.  Is there an easy way to test to see if a units schema is available in the current connection before the unitsProvider is set?

**Review by ColinKerr — APPROVED** [review]

**Comment by calebmshafer** [comment]
> @calebmshafer Can you give this a look when you get the chance? A couple groups a anxiously awaiting the units schema support on the frontend. Thanks!

@christophermlawson sure, I'll take a look today.

**Review by bsteinbk — APPROVED** [review]

**Comment by christophermlawson** [comment]
@calebmshafer I have the approvals I need. Do you want to have a look before I merge?

**Comment by calebmshafer** [comment]
@christophermlawson reviewing now...

File: core/quantity/src/Formatter/Format.ts
Comment (calebmshafer) [inline core/quantity/src/Formatter/Format.ts:185]: Why can't the Format class in ecschema-metadata directly use this format class rather having to create a base class to work with? I know it would have to wrap this one but not sure I'm seeing why we need two.

File: core/quantity/src/Formatter/Format.ts
Comment (calebmshafer) [inline core/quantity/src/Formatter/Format.ts:283]: Do we have an example of what is put into the CustomProps?

File: core/quantity/src/Interfaces.ts
Comment (calebmshafer) [inline core/quantity/src/Interfaces.ts:65]: Is this supposed to portray an error status?

File: full-stack-tests/core/src/backend/backend.ts
Comment (calebmshafer) [inline full-stack-tests/core/src/backend/backend.ts:125]: I _think_ we should be iterating over the list of Rpcs defined within the RpcInterfaces.ts file...

File: core/quantity/src/Formatter/Format.ts
Comment (bsteinbk) [reply core/quantity/src/Formatter/Format.ts:283]: CustomProps is not new - it was added to support custom QuantityTypes - see [Bearing example](https://github.com/iTwin/itwinjs-core/blob/5905cb1a48c4d790b5389d7f0ea141bc3ce95f23/test-apps/ui-test-app/src/frontend/api/BearingQuantityType.ts#L13)

File: full-stack-tests/core/src/backend/backend.ts
Comment (christophermlawson) [reply full-stack-tests/core/src/backend/backend.ts:125]: This took me the longest time to resolve.  It appears like all the other RPC implementations are registered explicitly (except the test ones) by IModelHost (line 394).  I didn't want to create a dependency to the rpc package in the backend, hence this decision. What do you think?

File: core/quantity/src/Interfaces.ts
Comment (christophermlawson) [reply core/quantity/src/Interfaces.ts:65]: not real familiar with this, but the FormatterSpec from the quantity package is setting this with a boolean if there is no input unit set (line 74).

File: core/quantity/src/Formatter/Format.ts
Comment (christophermlawson) [reply core/quantity/src/Formatter/Format.ts:283]: One example that I've seen is the BearingFormatProps interface in the BearingQuantityType ([https://github.com/iTwin/itwinjs-core/blob/master/core/frontend/src/test/BearingQuantityType.ts]).  It's whatever extra properties that a particular formatter implementation would need to format the quantity.

File: core/ecschema-rpc/common/package.json
Comment (christophermlawson) [reply core/ecschema-rpc/common/package.json]: Removed

File: core/frontend/src/quantity-formatting/QuantityFormatter.ts
Comment (christophermlawson) [reply core/frontend/src/quantity-formatting/QuantityFormatter.ts]: removed the comment

File: core/quantity/src/Interfaces.ts
Comment (bsteinbk) [reply core/quantity/src/Interfaces.ts:65]: @christophermlawson - this `error` boolean and its use in the FormatterSpec should not exist.  Neither are used in code on the master branch.

Comment (bsteinbk) [reply core/quantity/src/Interfaces.ts:65]: Basically all changes in FormatterSpec.ts should be reverted.

File: full-stack-tests/core/src/backend/backend.ts
Comment (calebmshafer) [reply full-stack-tests/core/src/backend/backend.ts:125]: Ah, I didn't realize we registered all of the others by default. I thought it was only a sub-set. If that's the case, this works with me. resolving.

File: core/quantity/src/Interfaces.ts
Comment (christophermlawson) [reply core/quantity/src/Interfaces.ts:65]: I've remove the prop and updated FormatterSpec

Comment (bsteinbk) [reply core/quantity/src/Interfaces.ts:65]: Did you push, because I don't see any changes yet?

Comment (christophermlawson) [reply core/quantity/src/Interfaces.ts]: sorry - should be there now.

File: core/backend/src/IModelSchemaLoader.ts
Comment (christophermlawson) [inline core/backend/src/IModelSchemaLoader.ts:5]: @calebmshafer @ColinKerr Here's what I came up with.  Note that I'm using 'import type' for Schema to so allow the getSchema call signatures to remain unchanged (type T would result in 'unknown' causing compilation errors in the calling code like in the IModelSchemaLoader.test.ts and IModelTransformer.test.ts).  I also updated the package.json to use peerDependenciesMeta.

File: core/backend/src/IModelSchemaLoader.ts
Comment (ColinKerr) [inline core/backend/src/IModelSchemaLoader.ts:50]: also list core quantity

File: core/backend/package.json
Comment (ColinKerr) [inline core/backend/package.json:47]: we should probably also list core quantity because that is needed if you include ECSchema-metadata

Comment (christophermlawson) [reply core/backend/package.json:47]: Not sure about this.  core-quantity should already be there once/if it's decided to install ecschema-metadata with the backend.  Meaning that if you install the backend you will have to install core-quantity.

File: core/backend/src/IModelSchemaLoader.ts
Comment (christophermlawson) [reply core/backend/src/IModelSchemaLoader.ts:53]: Right now ecschema-metadata has 2 peer dependencies or perhaps more in the future.  Do we want to list them all here?  The user will get a warning upon installation if they don't have the peers installed.

File: core/quantity/src/Formatter/Format.ts
Comment (christophermlawson) [reply core/quantity/src/Formatter/Format.ts:185]: I know we discussed this with Colin, so I will go ahead and resolve this conversation, if that's OK.

File: core/ecschema-metadata/package.json
Comment (christophermlawson) [reply core/ecschema-metadata/package.json]: Can you go ahead and resolve this?  ecschema-metadata is no optional as you know.

File: common/changes/@itwin/core-frontend/queryUnitsSchema_2022-02-01-19-53.json
Comment (calebmshafer) [inline common/changes/@itwin/core-frontend/queryUnitsSchema_2022-02-01-19-53.json:5]: Can we change this changelog message?

File: core/backend/src/IModelSchemaLoader.ts
Comment (ColinKerr) [reply core/backend/src/IModelSchemaLoader.ts:53]: OK

File: test-apps/ui-test-app/src/backend/main.ts
Comment (calebmshafer) [inline test-apps/ui-test-app/src/backend/main.ts:34]: Looks like some merge was messed up here as we lost the IModelsClient configured.

File: core/backend/package.json
Comment (calebmshafer) [reply core/backend/package.json:47]: Yea, I agree with Chris, I think unless it's needed without ecschema-metadata imported then it's not necessary to list it.

File: test-apps/ui-test-app/src/backend/main.ts
Comment (christophermlawson) [reply test-apps/ui-test-app/src/backend/main.ts:34]: Not sure I'm following.  Looking at master everything seems the same except for the registerImpl call I added.

File: common/changes/@itwin/core-frontend/queryUnitsSchema_2022-02-01-19-53.json
Comment (christophermlawson) [reply common/changes/@itwin/core-frontend/queryUnitsSchema_2022-02-01-19-53.json]: Done.  Thanks.

File: test-apps/ui-test-app/src/backend/main.ts
Comment (calebmshafer) [reply test-apps/ui-test-app/src/backend/main.ts:34]: Hm looks like `master` lost this change as well... Not sure what happened but I'll take a look.

Comment (calebmshafer) [reply test-apps/ui-test-app/src/backend/main.ts:34]: ignore that last message...

File: full-stack-tests/core/package.json
Comment (calebmshafer) [inline full-stack-tests/core/package.json:45]: nit: can you alphabetize?

File: core/frontend/src/quantity-formatting/UnitsData.ts
Comment (calebmshafer) [inline core/frontend/src/quantity-formatting/UnitsData.ts:12]: Is this used anywhere? @christophermlawson @bsteinbk

File: core/backend/package.json
Comment (calebmshafer) [reply core/backend/package.json:47]: @ColinKerr you okay with us resolving this?

**Review by calebmshafer — APPROVED** [review]
Changes look good. We will probably want to add something to the NextVersion.md to highlight the change. Not sure if you want to tie it in with what @bsteinbk is doing in the other PR.

**Comment by bsteinbk** [comment]
I have a follow-up PR that promotes the QuantityFormatter to public and adds learning documentation about this stuff.... I did not add anything to Next Version yet but I can in my PR.

**Comment by calebmshafer** [comment]
> I have a follow-up PR that promotes the QuantityFormatter to public and adds learning documentation about this stuff.... I did not add anything to Next Version yet but I can in my PR.

Awesome, thanks @bsteinbk!

**Pull request merged**
